diff -ruN a/vstrt/inference_helper.h b/vstrt/inference_helper.h
--- a/vstrt/inference_helper.h
+++ b/vstrt/inference_helper.h
@@ -7,7 +7,7 @@
 #include <string>
 #include <vector>
 
-#include <VSHelper.h>
+#include <VSHelper4.h>
 
 #include "cuda_helper.h"
 #include "trt_utils.h"
@@ -78,7 +78,7 @@
                         y * info.in.pitch + x * info.in.bytes_per_sample
                     };
 
-                    vs_bitblt(
+                    vsh::bitblt(
                         h_data, src_tile_w_bytes,
                         src_ptr, info.in.pitch,
                         static_cast<size_t>(src_tile_w_bytes),
@@ -110,7 +110,7 @@
                         info.h_scale * y * info.out.pitch + info.w_scale * x * info.out.bytes_per_sample
                     };
 
-                    vs_bitblt(
+                    vsh::bitblt(
                         dst_ptr + (y_crop_start * info.out.pitch + x_crop_start * info.out.bytes_per_sample),
                         info.out.pitch,
                         h_data + (y_crop_start * dst_tile_w_bytes + x_crop_start * info.out.bytes_per_sample),
diff -ruN a/vstrt/utils.h b/vstrt/utils.h
--- a/vstrt/utils.h
+++ b/vstrt/utils.h
@@ -11,8 +11,8 @@
 
 #include <NvInferRuntime.h>
 
-#include <VapourSynth.h>
-#include <VSHelper.h>
+#include <VapourSynth4.h>
+#include <VSHelper4.h>
 
 #ifdef __cpp_impl_reflection
 #include <meta>
@@ -49,16 +49,16 @@
     vi->width *= out_width / in_width;
 
     if (out_dims.d[1] == 1 || flexible_output) {
-        vi->format = vsapi->registerFormat(cmGray, sample_type, bits_per_sample, 0, 0, core);
+        vsapi->queryVideoFormat(&vi->format, cfGray, sample_type, bits_per_sample, 0, 0, core);
     } else if (out_dims.d[1] == 3) {
-        vi->format = vsapi->registerFormat(cmRGB, sample_type, bits_per_sample, 0, 0, core);
+        vsapi->queryVideoFormat(&vi->format, cfRGB, sample_type, bits_per_sample, 0, 0, core);
     }
 }
 
 static inline
 std::vector<const VSVideoInfo *> getVideoInfo(
     const VSAPI * vsapi,
-    const std::vector<VSNodeRef *> & nodes
+    const std::vector<VSNode *> & nodes
 ) noexcept {
 
     std::vector<const VSVideoInfo *> vis;
@@ -72,14 +72,14 @@
 }
 
 static inline
-std::vector<const VSFrameRef *> getFrames(
+std::vector<const VSFrame *> getFrames(
     int n,
     const VSAPI * vsapi,
     VSFrameContext * frameCtx,
-    const std::vector<VSNodeRef *> & nodes
+    const std::vector<VSNode *> & nodes
 ) noexcept {
 
-    std::vector<const VSFrameRef *> frames;
+    std::vector<const VSFrame *> frames;
     frames.reserve(std::size(nodes));
 
     for (const auto & node : nodes) {
@@ -95,7 +95,7 @@
 ) noexcept {
 
     for (const auto & vi : vis) {
-        if (!isConstantFormat(vi)) {
+        if (!vsh::isConstantVideoFormat(vi)) {
             return "video format must be constant";
         }
 
@@ -107,7 +107,7 @@
             return "number of frames mismatch";
         }
 
-        if (vi->format->subSamplingH != 0 || vi->format->subSamplingW != 0) {
+        if (vi->format.subSamplingH != 0 || vi->format.subSamplingW != 0) {
             return "clip must not be sub-sampled";
         }
     }
@@ -123,11 +123,11 @@
 ) noexcept {
 
     for (const auto & vi : vis) {
-        if (vi->format->sampleType != sample_type) {
+        if (vi->format.sampleType != sample_type) {
             return "sample type mismatch";
         }
 
-        if (vi->format->bitsPerSample != bits_per_sample) {
+        if (vi->format.bitsPerSample != bits_per_sample) {
             return "bits per sample mismatch";
         }
     }
@@ -143,7 +143,7 @@
     int num_planes = 0;
 
     for (const auto & vi : vis) {
-        num_planes += vi->format->numPlanes;
+        num_planes += vi->format.numPlanes;
     }
 
     return num_planes;
@@ -186,29 +186,29 @@
 ) {
 
     int err;
-    int device_id = static_cast<int>(vsapi->propGetInt(in, "device_id", 0, &err));
+    int device_id = static_cast<int>(vsapi->mapGetInt(in, "device_id", 0, &err));
     if (err) {
         device_id = 0;
     }
 
     cudaDeviceProp prop;
     if (auto error = cudaGetDeviceProperties(&prop, device_id); error != cudaSuccess) {
-        vsapi->setError(out, cudaGetErrorString(error));
+        vsapi->mapSetError(out, cudaGetErrorString(error));
         return ;
     }
 
     auto setProp = [&](const char * name, const auto & value, int data_length = -1) {
         using T = std::decay_t<decltype(value)>;
         if constexpr (std::is_integral_v<T>) {
-            vsapi->propSetInt(out, name, static_cast<int64_t>(value), paReplace);
+            vsapi->mapSetInt(out, name, static_cast<int64_t>(value), maReplace);
         } else if constexpr (std::is_same_v<T, const char *>) {
-            vsapi->propSetData(out, name, value, data_length, paReplace);
+            vsapi->mapSetData(out, name, value, data_length, dtBinary, maReplace);
         } else if constexpr (std::is_integral_v<std::remove_pointer_t<T>>) {
             std::array<int64_t, std::extent_v<std::remove_reference_t<decltype(value)>>> data;
             for (int i = 0; i < static_cast<int>(std::size(data)); i++) {
                 data[i] = value[i];
             }
-            vsapi->propSetIntArray(out, name, std::data(data), static_cast<int>(std::size(data)));
+            vsapi->mapSetIntArray(out, name, std::data(data), static_cast<int>(std::size(data)));
         }
     };
 
@@ -226,7 +226,7 @@
             for (int i = 0; i < 16; ++i) {
                 uuid[i] = prop.uuid.bytes[i];
             }
-            vsapi->propSetIntArray(out, "uuid", std::data(uuid), static_cast<int>(std::size(uuid)));
+            vsapi->mapSetIntArray(out, "uuid", std::data(uuid), static_cast<int>(std::size(uuid)));
         } else if constexpr (identifier_of(r) != "reserved") {
             setProp(std::string(identifier_of(r)).c_str(), prop.[:r:]);
         }
@@ -238,7 +238,7 @@
         for (int i = 0; i < 16; ++i) {
             uuid[i] = prop.uuid.bytes[i];
         }
-        vsapi->propSetIntArray(out, "uuid", std::data(uuid), static_cast<int>(std::size(uuid)));
+        vsapi->mapSetIntArray(out, "uuid", std::data(uuid), static_cast<int>(std::size(uuid)));
     }
     setProp("total_global_memory", prop.totalGlobalMem);
     setProp("shared_memory_per_block", prop.sharedMemPerBlock);
diff -ruN a/vstrt/vs_tensorrt.cpp b/vstrt/vs_tensorrt.cpp
--- a/vstrt/vs_tensorrt.cpp
+++ b/vstrt/vs_tensorrt.cpp
@@ -12,8 +12,8 @@
 #include <variant>
 #include <vector>
 
-#include <VapourSynth.h>
-#include <VSHelper.h>
+#include <VapourSynth4.h>
+#include <VSHelper4.h>
 
 #include <cuda_runtime.h>
 #include <NvInferRuntime.h>
@@ -42,7 +42,7 @@
 
 using namespace std::string_literals;
 
-static const VSPlugin * myself = nullptr;
+static VSPlugin * myself = nullptr;
 
 struct TicketSemaphore {
     std::atomic<intptr_t> ticket {};
@@ -72,7 +72,7 @@
 std::unique_ptr<Logger> logger;
 
 struct vsTrtData {
-    std::vector<VSNodeRef *> nodes;
+    std::vector<VSNode *> nodes;
     std::unique_ptr<VSVideoInfo> out_vi;
 
     int device_id;
@@ -111,30 +111,17 @@
     }
 };
 
-static void VS_CC vsTrtInit(
-    VSMap *in,
-    VSMap *out,
-    void **instanceData,
-    VSNode *node,
-    VSCore *core,
-    const VSAPI *vsapi
-) noexcept {
-
-    auto d = static_cast<vsTrtData *>(*instanceData);
-    vsapi->setVideoInfo(d->out_vi.get(), 1, node);
-}
-
-static const VSFrameRef *VS_CC vsTrtGetFrame(
+static const VSFrame *VS_CC vsTrtGetFrame(
     int n,
     int activationReason,
-    void **instanceData,
+    void *instanceData,
     void **frameData,
     VSFrameContext *frameCtx,
     VSCore *core,
     const VSAPI *vsapi
 ) noexcept {
 
-    auto d = static_cast<vsTrtData *>(*instanceData);
+    auto d = static_cast<vsTrtData *>(instanceData);
 
     if (activationReason == arInitial) {
         for (const auto & node : d->nodes) {
@@ -145,7 +132,7 @@
             getVideoInfo(vsapi, d->nodes)
         };
 
-        const std::vector<const VSFrameRef *> src_frames {
+        const std::vector<const VSFrame *> src_frames {
             getFrames(n, vsapi, frameCtx, d->nodes)
         };
 
@@ -166,17 +153,17 @@
         std::vector<const uint8_t *> src_ptrs;
         src_ptrs.reserve(src_planes);
         for (int i = 0; i < std::ssize(d->nodes); ++i) {
-            for (int j = 0; j < in_vis[i]->format->numPlanes; ++j) {
+            for (int j = 0; j < in_vis[i]->format.numPlanes; ++j) {
                 src_ptrs.emplace_back(vsapi->getReadPtr(src_frames[i], j));
             }
         }
 
-        VSFrameRef * const dst_frame { vsapi->newVideoFrame(
-            d->out_vi->format, d->out_vi->width, d->out_vi->height,
+        VSFrame * const dst_frame { vsapi->newVideoFrame(
+            &d->out_vi->format, d->out_vi->width, d->out_vi->height,
             src_frames[0], core
         )};
 
-        std::vector<VSFrameRef *> dst_frames;
+        std::vector<VSFrame *> dst_frames;
 
 #if NV_TENSORRT_MAJOR * 100 + NV_TENSORRT_MINOR >= 805 || defined(TRT_MAJOR_RTX)
         auto output_name = d->engines[0]->getIOTensorName(1);
@@ -198,7 +185,7 @@
         } else {
             for (int i = 0; i < dst_planes; ++i) {
                 auto frame { vsapi->newVideoFrame(
-                    d->out_vi->format, d->out_vi->width, d->out_vi->height,
+                    &d->out_vi->format, d->out_vi->width, d->out_vi->height,
                     src_frames[0], core
                 )};
                 dst_frames.emplace_back(frame);
@@ -213,14 +200,14 @@
             .in = InputInfo {
                 .width = vsapi->getFrameWidth(src_frames[0], 0),
                 .height = vsapi->getFrameHeight(src_frames[0], 0),
-                .pitch = vsapi->getStride(src_frames[0], 0),
-                .bytes_per_sample = vsapi->getFrameFormat(src_frames[0])->bytesPerSample,
+                .pitch = static_cast<int>(vsapi->getStride(src_frames[0], 0)),
+                .bytes_per_sample = vsapi->getVideoFrameFormat(src_frames[0])->bytesPerSample,
                 .tile_w = src_tile_w,
                 .tile_h = src_tile_h
             },
             .out = OutputInfo {
-                .pitch = vsapi->getStride(dst_frame, 0),
-                .bytes_per_sample = vsapi->getFrameFormat(dst_frame)->bytesPerSample
+                .pitch = static_cast<int>(vsapi->getStride(dst_frame, 0)),
+                .bytes_per_sample = vsapi->getVideoFrameFormat(dst_frame)->bytesPerSample
             },
             .w_scale = w_scale,
             .h_scale = h_scale,
@@ -256,11 +243,11 @@
         }
         
         if (!d->flexible_output_prop.empty()) {
-            auto prop = vsapi->getFramePropsRW(dst_frame);
+            auto prop = vsapi->getFramePropertiesRW(dst_frame);
 
             for (int i = 0; i < dst_planes; i++) {
                 auto key { d->flexible_output_prop + std::to_string(i) };
-                vsapi->propSetFrame(prop, key.c_str(), dst_frames[i], paReplace);
+                vsapi->mapSetFrame(prop, key.c_str(), dst_frames[i], maReplace);
                 vsapi->freeFrame(dst_frames[i]);
             }
         }
@@ -293,20 +280,20 @@
 
     auto d { std::make_unique<vsTrtData>() };
 
-    int num_nodes = vsapi->propNumElements(in, "clips");
+    int num_nodes = vsapi->mapNumElements(in, "clips");
     d->nodes.reserve(num_nodes);
     for (int i = 0; i < num_nodes; ++i) {
-        d->nodes.emplace_back(vsapi->propGetNode(in, "clips", i, nullptr));
+        d->nodes.emplace_back(vsapi->mapGetNode(in, "clips", i, nullptr));
     }
 
     auto set_error = [&](const std::string & error_message) {
-        vsapi->setError(out, (__func__ + ": "s + error_message).c_str());
+        vsapi->mapSetError(out, (__func__ + ": "s + error_message).c_str());
         for (const auto & node : d->nodes) {
             vsapi->freeNode(node);
         }
     };
 
-    const char * engine_path = vsapi->propGetData(in, "engine_path", 0, nullptr);
+    const char * engine_path = vsapi->mapGetData(in, "engine_path", 0, nullptr);
 
     std::vector<const VSVideoInfo *> in_vis;
     in_vis.reserve(std::size(d->nodes));
@@ -318,8 +305,8 @@
     }
 
     int error1, error2;
-    d->overlap_w = int64ToIntS(vsapi->propGetInt(in, "overlap", 0, &error1));
-    d->overlap_h = int64ToIntS(vsapi->propGetInt(in, "overlap", 1, &error2));
+    d->overlap_w = vsh::int64ToIntS(vsapi->mapGetInt(in, "overlap", 0, &error1));
+    d->overlap_h = vsh::int64ToIntS(vsapi->mapGetInt(in, "overlap", 1, &error2));
     if (!error1) {
         if (error2) {
             d->overlap_h = d->overlap_w;
@@ -333,8 +320,8 @@
         d->overlap_h = 0;
     }
 
-    int tile_w = int64ToIntS(vsapi->propGetInt(in, "tilesize", 0, &error1));
-    int tile_h = int64ToIntS(vsapi->propGetInt(in, "tilesize", 1, &error2));
+    int tile_w = vsh::int64ToIntS(vsapi->mapGetInt(in, "tilesize", 0, &error1));
+    int tile_h = vsh::int64ToIntS(vsapi->mapGetInt(in, "tilesize", 1, &error2));
 
     TileSize tile_size;
     if (!error1) { // manual specification triggered
@@ -370,7 +357,7 @@
 
     int error;
 
-    int device_id = int64ToIntS(vsapi->propGetInt(in, "device_id", 0, &error));
+    int device_id = vsh::int64ToIntS(vsapi->mapGetInt(in, "device_id", 0, &error));
     if (error) {
         device_id = 0;
     }
@@ -384,17 +371,17 @@
     }
     d->device_id = device_id;
 
-    d->use_cuda_graph = !!vsapi->propGetInt(in, "use_cuda_graph", 0, &error);
+    d->use_cuda_graph = !!vsapi->mapGetInt(in, "use_cuda_graph", 0, &error);
     if (error) {
         d->use_cuda_graph = false;
     }
 
-    d->num_streams = int64ToIntS(vsapi->propGetInt(in, "num_streams", 0, &error));
+    d->num_streams = vsh::int64ToIntS(vsapi->mapGetInt(in, "num_streams", 0, &error));
     if (error) {
         d->num_streams = 1;
     }
 
-    int verbosity = int64ToIntS(vsapi->propGetInt(in, "verbosity", 0, &error));
+    int verbosity = vsh::int64ToIntS(vsapi->mapGetInt(in, "verbosity", 0, &error));
     if (error) {
         verbosity = int(nvinfer1::ILogger::Severity::kWARNING);
     }
@@ -403,7 +390,7 @@
     }
     logger->set_verbosity(static_cast<nvinfer1::ILogger::Severity>(verbosity));
 
-    auto flexible_output_prop = vsapi->propGetData(in, "flexible_output_prop", 0, &error);
+    auto flexible_output_prop = vsapi->mapGetData(in, "flexible_output_prop", 0, &error);
     if (!error) {
         d->flexible_output_prop = flexible_output_prop;
     }
@@ -412,7 +399,7 @@
     // related to https://github.com/AmusementClub/vs-mlrt/discussions/65, for unknown reason
 #if !(NV_TENSORRT_MAJOR == 9 && defined(_WIN32)) && !defined(TRT_MAJOR_RTX)
     if (!initLibNvInferPlugins(logger.get(), "")) {
-        vsapi->logMessage(mtWarning, "vstrt: Initialize TensorRT plugins failed");
+        vsapi->logMessage(mtWarning, "vstrt: Initialize TensorRT plugins failed", core);
     }
 #endif
 #endif
@@ -446,7 +433,7 @@
         uint64_t diagnostics;
         auto engine_validity = d->runtime->getEngineValidity(engine_data.get(), static_cast<int64_t>(engine_nbytes), &diagnostics);
         if (engine_validity == nvinfer1::EngineValidity::kSUBOPTIMAL) {
-            vsapi->logMessage(mtWarning, "suboptimal engine");
+            vsapi->logMessage(mtWarning, "suboptimal engine", core);
         } else if (engine_validity == nvinfer1::EngineValidity::kINVALID) {
             std::ostringstream diagnostics_message;
             diagnostics_message << "invalid engine: ";
@@ -606,23 +593,27 @@
         #else // NV_TENSORRT_MAJOR * 100 + NV_TENSORRT_MINOR >= 805 || defined(TRT_MAJOR_RTX)
             const nvinfer1::Dims & out_dims = exec_context->getBindingDimensions(1);
         #endif // NV_TENSORRT_MAJOR * 100 + NV_TENSORRT_MINOR >= 805 || defined(TRT_MAJOR_RTX)
-        vsapi->propSetInt(out, "num_planes", out_dims.d[1], paReplace);
+        vsapi->mapSetInt(out, "num_planes", out_dims.d[1], maReplace);
     }
 
-    vsapi->createFilter(
-        in, out, "Model",
-        vsTrtInit, vsTrtGetFrame, vsTrtFree,
-        fmParallel, 0, d.release(), core
+    std::vector<VSFilterDependency> deps;
+    for (const auto & node : d->nodes) {
+        deps.push_back({node, rpStrictSpatial});
+    }
+    auto out_vi = d->out_vi.get();
+    vsapi->createVideoFilter(
+        out, "Model", out_vi,
+        vsTrtGetFrame, vsTrtFree,
+        fmParallel, deps.data(), static_cast<int>(deps.size()), d.release(), core
     );
 }
 
-VS_EXTERNAL_API(void) VapourSynthPluginInit(
-    VSConfigPlugin configFunc,
-    VSRegisterFunction registerFunc,
-    VSPlugin *plugin
+VS_EXTERNAL_API(void) VapourSynthPluginInit2(
+    VSPlugin *plugin,
+    const VSPLUGINAPI *vspapi
 ) noexcept {
 
-    configFunc(
+    vspapi->configPlugin(
 #if defined(TRT_MAJOR_RTX)
         "io.github.amusementclub.vs_tensorrt_rtx",
         "trt_rtx", "TensorRT-RTX ML Filter Runtime",
@@ -630,7 +621,7 @@
         "io.github.amusementclub.vs_tensorrt",
         "trt", "TensorRT ML Filter Runtime",
 #endif
-        VAPOURSYNTH_API_VERSION, 1, plugin
+        VS_MAKE_VERSION(1, 0), VAPOURSYNTH_API_VERSION, 0, plugin
     );
 
     // TRT 9 for windows does not export getInferLibVersion()
@@ -684,8 +675,8 @@
 
     myself = plugin;
 
-    registerFunc("Model",
-        "clips:clip[];"
+    vspapi->registerFunction("Model",
+        "clips:vnode[];"
         "engine_path:data;"
         "overlap:int[]:opt;"
         "tilesize:int[]:opt;"
@@ -694,44 +685,45 @@
         "num_streams:int:opt;"
         "verbosity:int:opt;"
         "flexible_output_prop:data:opt;",
+        "clip:vnode;",
         vsTrtCreate,
         nullptr,
         plugin
     );
 
     auto getVersion = [](const VSMap *, VSMap * out, void *, VSCore *, const VSAPI *vsapi) {
-        vsapi->propSetData(out, "version", VERSION, -1, paReplace);
+        vsapi->mapSetData(out, "version", VERSION, -1, dtBinary, maReplace);
 
-        vsapi->propSetData(
+        vsapi->mapSetData(
             out, "tensorrt_version",
 #if NV_TENSORRT_MAJOR == 9 && defined(_WIN32) && !defined(TRT_MAJOR_RTX)
             std::to_string(NV_TENSORRT_VERSION).c_str(), 
 #else
             std::to_string(getInferLibVersion()).c_str(), 
 #endif
-            -1, paReplace
+            -1, dtBinary, maReplace
         );
 
-        vsapi->propSetData(
+        vsapi->mapSetData(
             out, "tensorrt_version_build",
-            std::to_string(NV_TENSORRT_VERSION).c_str(), -1, paReplace
+            std::to_string(NV_TENSORRT_VERSION).c_str(), -1, dtBinary, maReplace
         );
 
         int runtime_version;
         cudaRuntimeGetVersion(&runtime_version);
-        vsapi->propSetData(
+        vsapi->mapSetData(
             out, "cuda_runtime_version",
-            std::to_string(runtime_version).c_str(), -1, paReplace
+            std::to_string(runtime_version).c_str(), -1, dtBinary, maReplace
         );
 
-        vsapi->propSetData(
+        vsapi->mapSetData(
             out, "cuda_runtime_version_build",
-            std::to_string(__CUDART_API_VERSION).c_str(), -1, paReplace
+            std::to_string(__CUDART_API_VERSION).c_str(), -1, dtBinary, maReplace
         );
 
-        vsapi->propSetData(out, "path", vsapi->getPluginPath(myself), -1, paReplace);
+        vsapi->mapSetData(out, "path", vsapi->getPluginPath(myself), -1, dtBinary, maReplace);
     };
-    registerFunc("Version", "", getVersion, nullptr, plugin);
+    vspapi->registerFunction("Version", "", "any", getVersion, nullptr, plugin);
 
-    registerFunc("DeviceProperties", "device_id:int:opt;", getDeviceProp, nullptr, plugin);
+    vspapi->registerFunction("DeviceProperties", "device_id:int:opt;", "any", getDeviceProp, nullptr, plugin);
 }
